### ISSUE TITLE
Starts cloning, shows progress, but then finishes without anything there.

### ISSUE THREAD

**Issue by maks** [maks/MGit#54]
_From @johann-petrak on November 12, 2015 0:1_

Just installed SGit on a Samsung Note Pro 12.2 (Model SM-P900) with Android 4.4.2.
I have tried to install (rather large) git repositories using SSH (the private key is not password protected) and https with userid and password and both cases, the cloning seems to start, SGit shows progress but then the entry disappears from the list and nothing has been cloned. 
None of my repositories can be cloned, I made several attempts for each. No error message is shown in the window. 

_Copied from original issue: sheimi/SGit#169_

**Comment by maks** [comment]
_From @Lrs121 on November 12, 2015 0:14_

I have noticed this occurs with large(ish) repos. I have a few theories as to why it happens but i havent confirmed any of them yet. Cloning smaller reops doesnt seem to cause the issue. it may not be possible to clone large repos at this time.

**Comment by maks** [comment]
_From @FabianInostroza on November 12, 2015 4:18_

I just today noticed this problem too, at the end, the following message is shown: "attempt to use Inflater after calling end".

**Comment by maks** [comment]
_From @Lrs121 on November 12, 2015 5:23_

Thats what i remember. though i have no idea what that means

**Comment by maks** [comment]
Large repos are likely causing out of memory issues due to the limited heap size available to apps, especially on a 4.4 devices as max heap size is in general lower on earlier versions of Android.
It would be very helpful if someone on this thread could post a URL for a public repo where they see this error occurring so I can  confirm this is the root cause.

**Comment by maks** [comment]
_From @Lrs121 on May 21, 2016 5:5_

I don't remember which repo was specifically causing me problems but I believe the git for the Linux kernel might cause it. I'll take a look around in the morning for one that will definitely cause it and link it here.

**Comment by maks** [comment]
@Lrs121 yikes, the kernel repo would definitely do it.
The memory usage is due to the need to inflate objects inside pack files, which are "deflated" using zlib (see: https://www.kernel.org/pub/software/scm/git/docs/technical/pack-format.txt and https://codewords.recurse.com/issues/three/unpacking-git-packfiles).

 Jgit (the git library that SGit app uses) has a "streaming" mode that should allow it to cope with large compressed objects in packfiles, up to a point - I'll check how that paramter is set in the current SGit release...
This has been quite well documented in the egit bugtracker (egit being the original client for the jgit library):
https://bugs.eclipse.org/bugs/show_bug.cgi?id=394078

**Comment by maks** [comment]
_From @Lrs121 on May 21, 2016 20:23_

With understanding how that works I believe the stream gets overloaded with the current setup, for one reason or another. I do know that android has a flag for large memory allocations, I don't remember if that's used currently. If it is not, perhaps if that is used in conjunction with the streaming mode will help handle slightly larger repos. It probably wouldn't be able to hand the kernel repo, but it may give other repos a fighting chance of succeeding. 

Another option maybe to include a clone depth setting that allows users to only pull a specific number of recent commits. Much like clone-depth=1. I don't know if Jgit has that implemented. If it does it may help with some overhead.

**Comment by maks** [comment]
The comment https://github.com/sheimi/SGit/issues/65#issuecomment-73890069 from #65 is worth looking into as Agit of cocurse also uses jgit.

**Comment by maks** [comment]
_From @Maurdekye on August 16, 2016 5:4_

After looking into this problem and coming upon this issue thread, I'm confused because it seems to be happening with my repo as well, even though it's only about 7.7mb. I don't think that could be considered a large repo. I'm currently running rooted kitkat 5.1 on an LG G4.

**Comment by maks** [comment]
@Maurdekye If you have a public repo that is consistently showing this behaviour, that would be a big help in debugging and testing the possible fix.

**Comment by leighmcculloch** [comment]
I'm experiencing this issue also. It gets to 100% when cloning then disappears.

**Comment by spacekitteh** [comment]
Is there any update on this?

**Comment by maks** [comment]
@spacekitteh are you seeing a simliar problem?

**Comment by spacekitteh** [comment]
Yep, I added a ~60mb pdf to my git repo. When I pulled on my Galaxy Tab S7, it'd fail.

**Comment by maks** [comment]
Ah ok, at a guess I would think its due to the large size of the repo. JGit, the git library that MGit makes use of has issues on Android with large repos or repos that have large files, due to not using streaming, so all the git operations on the repo contents needs to happen in ram and depending on your device, [this comment ](https://github.com/maks/MGit/issues/459#issuecomment-510250495)has more info and links to even more details.
I'd like to be able to address this at some stage but for now I'll try to at least get better error reporting for this done the next chance I get.

**Comment by oliviercarizzoni** [comment]
I don't know if you are still working on this, but I am having the same problem with a large repo (700Mo) with numerous commits, is there something to do ?

**Comment by maks** [comment]
@oliviercarizzoni  Sadly at a bit of an impasse on this issue. From memory and per previous comments, there was a known issue with streaming data with JGit which was fixed in JGit but MGit is unable to move to newer versions of JGit due to its use of Java APIs not available on older versions of Android.

**Comment by AidosZhaqyp** [comment]
> @oliviercarizzoni  Sadly at a bit of an impasse on this issue. From memory and per previous comments, there was a known issue with streaming data with JGit which was fixed in JGit but MGit is unable to move to newer versions of JGit due to its use of Java APIs not available on older versions of Android.

  Hi. I stii have that problem and it's really annoying. You can't update JGit because new versions use Java nio and it supports starting Android 8? Could you please tell me in which version of JGit it was fixed? And what about the IDE to have 2 versions of app? One version with current version of JGit and another one with newer JGit?

**Comment by 0dminnimda** [comment]
Hi, I am also having this issue with realtively small repo https://github.com/mephi-tex/mephi-tex/tree/45058bf6ee1fdc66af8047fcc96ffa56b1fec022
Is there any workaround?

**Comment by 0dminnimda** [comment]
As @oliviercarizzoni proposed, I cloned the repo via termux and imported it, and it wotks fine

**Comment by maks** [comment]
@0dminnimda I see you edited your original comment, so just to clarify, I would not consider a 400MB repo  "realtively small" [sic] from the point of view of a Android app where *historically* max heap sizes available to Java (not NDK) based apps was as little as 16MB (https://android-doc.github.io/reference/android/app/ActivityManager.html#getMemoryClass()) and is completely dependant on the Android device in use. More modern devices will have higher heap size limits but even eg. a 512MB limit could be insufficient for a repo of this size due to other overheads in heap usage in the app.
